1.9 release docs (#329)
kubefirst/kubefirst: docs/tooling/kubefirst-cli.md

# Kubefirst Command Line

Kubefirst provides a CLI to empower the whole previsioning cluster management process. The basic usage of the CLI is to

> fully create the Kubefirst previsioning process, but there are also many other handful options to backup, validation and

Reviewer: ```suggestion
fully create the Kubefirst provisioning process, but there are also many other helpful options to backup, validate and
```

> cluster destroy.

Reviewer: ```suggestion
destroy.
```

> ## Kubefirst CLI available commands

Reviewer: ```suggestion
## Kubefirst CLI Available Commands
```

- [clean](#clean)
- [init](#init)
- [cluster](#cluster-create)
- [destroy](#destroy)
- [backupSSL](#backupssl)
- [restoreSSL](#restoressl)
- [state](#state)

### clean

Kubefirst creates files, folders and cloud buckets during installation at your environment. This command removes and

> re-create Kubefirst base files. To destroy cloud resources you need to specify additional flags (--destroy-buckets)

Reviewer: ```suggestion
re-creates Kubefirst base files. To destroy cloud resources you need to specify additional flags (--destroy-buckets)
```

```
Usage:
kubefirst clean [flags]

Flags:
--destroy-buckets   destroy buckets created by init cmd
--destroy-confirm   confirm destroy operation (to be used during automation process)
-h, --help              help for clean
--rm-logs           remove logs folder
```

### init

> Init command will prepare the installation, and create the initial resources that will support the installation when 

Reviewer: ```suggestion
The **init** command will prepare the installation and create the initial resources that will support the installation when 
```

> `kubefirst cluster create` is be called.

Reviewer: ```suggestion
`kubefirst cluster create` is called.
```

```
Usage:
kubefirst init [flags]

Flags:

> --admin-email string        the email address for the administrator as well as for lets-encrypt certificate emails

Reviewer: ```suggestion
--admin-email string        The email address for the administrator as well as for Lets-Encrypt certificate emails.
```

> --aws-assume-role string    instead of using AWS IAM user credentials, AWS AssumeRole feature generate role based credentials, more at https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html

Reviewer: ```suggestion
--aws-assume-role string    Instead of using AWS IAM user credentials, AWS AssumeRole feature generates role based credentials, more at https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html.
```

> --aws-nodes-spot            nodes spot on AWS EKS compute nodes

Reviewer: ```suggestion
--aws-nodes-spot            Nodes spot on AWS EKS compute nodes.
```

> --clean                     delete any local kubefirst content ~/.kubefirst, ~/.k1

Reviewer: ```suggestion
--clean                     Delete any local Kubefirst content ~/.kubefirst, ~/.k1.
```

> --cloud string              the cloud to provision infrastructure in

Reviewer: ```suggestion
--cloud string              The cloud in which to provision infrastructure.
```

> --cluster-name string       the cluster name, used to identify resources on cloud provider (default "kubefirst")

Reviewer: ```suggestion
--cluster-name string       The cluster name, used to identify resources on cloud provider (default "kubefirst").
```

> --dry-run                   set to dry-run mode, no changes done on cloud provider selected

Reviewer: ```suggestion
--dry-run                   Set to dry-run mode - no changes done on cloud provider selected.
```

--github-host string        Github URL (default "github.com")
--github-org string         Github Org of repos
--github-owner string       Github owner of repos
--github-user string        Github user

> --gitops-branch string      version/branch used on git clone - former: version-gitops flag

Reviewer: ```suggestion
--gitops-branch string      Version/branch used on Git clone - former: version-gitops flag.
```

> --gitops-owner string       git owner of gitops, this may be a user or a org to support forks for testing (default "kubefirst")

Reviewer: ```suggestion
--gitops-owner string       Git owner of GitOps - this may be a user or an org to support forks for testing (default "kubefirst").
```

> --gitops-repo string        version/branch used on git clone (default "gitops")

Reviewer: ```suggestion
--gitops-repo string        Version/branch used on Git clone (default "gitops").
```

> -h, --help                      help for init

Reviewer: ```suggestion
-h, --help                   Help for init.
```

> --hosted-zone-name string   the domain to provision the kubefirst platform in

Reviewer: ```suggestion
--hosted-zone-name string   The domain in which to provision the Kubefirst platform.
```

--metaphor-branch string    version/branch used on git clone - former: version-gitops flag

> --profile string            AWS profile located at ~/.aws/config (default "default")

Reviewer: ```suggestion
--profile string            AWS profile located at ~/.aws/config (default "default").
```

> --region string             the region to provision the cloud resources in (default "eu-west-1")

Reviewer: ```suggestion
--region string             The region in which to provision the cloud resources (default "eu-west-1").
```

> --s3-suffix string          unique identifier for s3 buckets

Reviewer: ```suggestion
--s3-suffix string          Unique identifier for S3 buckets.
```

> --template-tag string       fallback tag used on git clone.

Reviewer: ```suggestion
--template-tag string       Fallback tag used on Git clone.
```

> Details: if "gitops-branch" is provided, branch("gitops-branch") has precedence and installer will attempt to clone branch("gitops-branch") first,

Reviewer: ```suggestion
Details: If "gitops-branch" is provided, branch("gitops-branch") has precedence and the installer will attempt to clone the branch("gitops-branch") first,
```

if it fails, then fallback it will attempt to clone the tag provided at "template-tag" flag (default "1.8.6")

> --use-telemetry             installer will not send telemetry about this installation (default true)

Reviewer: ```suggestion
--use-telemetry             Installer will not send telemetry about this installation (default true).
```

```

### cluster create

> Cluster Level operations like create a new cluster provisioned with all kubefirst goodies.

Reviewer: ```suggestion
Cluster Level operations like **create a new cluster** provisioned with all Kubefirst goodies.
```

```
Usage:
kubefirst cluster [flags]
kubefirst cluster [command]

Available Commands:

> create        create a kubefirst management cluster

Reviewer: ```suggestion
create        Create a Kubefirst management cluster.
```

> create-github create a kubefirst management cluster with github as Git Repo

Reviewer: ```suggestion
create-github Create a Kubefirst management cluster with GitHub as Git Repo.
```

> create-gitlab create a kubefirst management cluster

Reviewer: ```suggestion
create-gitlab Create a Kubefirst management cluster.
```

Flags:

> -h, --help   help for cluster

Reviewer: ```suggestion
-h, --help   Help for cluster.
```

Use "kubefirst cluster [command] --help" for more information about a command.
```

### destroy

> Destroy the kubefirst management cluster and all of the components in kubernetes.

Reviewer: ```suggestion
Destroy the Kubefirst management cluster and all of the components in Kubernetes.
```

> Optional: skip gitlab terraform if the registry has already been deleted.

Reviewer: ```suggestion
Optional: Skip GitLab Terraform if the registry has already been deleted.
```

```
Usage:
kubefirst destroy [flags]

Flags:

> --dry-run                 set to dry-run mode, no changes done on cloud provider selected

Reviewer: ```suggestion
--dry-run                 Set to dry-run mode, no changes done on cloud provider selected.
```

> -h, --help                    help for destroy

Reviewer: ```suggestion
-h, --help                Help for destroy.
```

> --skip-base-terraform     whether to skip the terraform destroy against base install - note: if you already deleted registry it doesnt exist

Reviewer: ```suggestion
--skip-base-terraform     Whether to skip the Terraform destroy against the base install. Note: If you already deleted the registry, it doesn't exist.
```

> --skip-delete-register    whether to skip deletion of register application

Reviewer: ```suggestion
--skip-delete-register    Whether to skip the deletion of register application.
```

> --skip-gitlab-terraform   whether to skip the terraform destroy against gitlab - note: if you already deleted registry it doesnt exist

Reviewer: ```suggestion
--skip-gitlab-terraform   Whether to skip the Terraform destroy against GitLab. Note: If you already deleted the registry, it doesn't exist.
```

```

### backupSSL

> This command create a backup of secrets from cert manager certificates to bucket named kubefirst-<DOMAIN> where can be 

Reviewer: ```suggestion
This command creates a backup of secrets from cert manager certificates to a bucket named Kubefirst-<DOMAIN> where secrets can be 
```

> used on provisioning phase with the flag --recycle-ssl

Reviewer: ```suggestion
used on the provisioning phase with the flag --recycle-ssl.
```

```
Usage:
kubefirst backupSSL [flags]

Flags:

> -h, --help   help for backupSSL

Reviewer: ```suggestion
-h, --help   Help for backupSSL.
```

```

### restoreSSL

> Restore the backup SSL that is stored in S3 bucket, and avoid re-issuing new certificates.

Reviewer: ```suggestion
Restore the backup SSL that is stored in the S3 bucket and avoid re-issuing new certificates.
```

```
Usage:
kubefirst restoreSSL [flags]

Flags:

> --dry-run         set to dry-run mode, no changes done on cloud provider selected

Reviewer: ```suggestion
--dry-run         Set to dry-run mode, no changes done on cloud provider selected.
```

> -h, --help            help for restoreSSL

Reviewer: ```suggestion
-h, --help        Help for restoreSSL.
```

> --use-telemetry   installer will not send telemetry about this installation (default true)

Reviewer: ```suggestion
--use-telemetry   Installer will not send telemetry about this installation (default true).
```

```

### state

Kubefirst configuration can be handed over to another user by pushing the Kubefirst config files to a S3 bucket.

```
Usage:
kubefirst state [flags]

Flags:

> --bucket-name string   set the bucket name to store the Kubefirst config file

Reviewer: ```suggestion
--bucket-name string   Set the bucket name to store the Kubefirst config file.
```

> -h, --help                 help for state

Reviewer: ```suggestion
-h, --help             Help for state.
```

> --pull                 pull Kubefirst config file to the S3 bucket

Reviewer: ```suggestion
--pull                 Pull the Kubefirst config file to the S3 bucket.
```

> --push                 push Kubefirst config file to the S3 bucket

Reviewer: ```suggestion
--push                 Push Kubefirst config file to the S3 bucket.
```

> --region string        set S3 bucket region

Reviewer: ```suggestion
--region string        Set S3 bucket region.
```
